api/libs/token.py

```diff
@@ -0,0 +1,187 @@
+        return request.headers.get(HEADER_NAME_PASSPORT)
+
+    ret = _try_extract_passport_token_from_cookie(request) or _try_extract_passport_token_from_header(request)
+    return ret
+
+
+def set_access_token_to_cookie(request: Request, response: Response, token: str, samesite: str = "Lax"):
+    response.set_cookie(
+        COOKIE_NAME_ACCESS_TOKEN,
+        value=token,
+        httponly=True,
+        secure=is_secure(),
+        samesite=samesite,
+        max_age=int(dify_config.ACCESS_TOKEN_EXPIRE_MINUTES * 60),
+        path="/",
+    )
+
+
+def set_refresh_token_to_cookie(request: Request, response: Response, token: str):
+    response.set_cookie(
+        COOKIE_NAME_REFRESH_TOKEN,
+        value=token,
+        httponly=True,
+        secure=is_secure(),
+        samesite="Lax",
+        max_age=int(60 * 60 * 24 * dify_config.REFRESH_TOKEN_EXPIRE_DAYS),
+        path="/",
+    )
+
+
+def set_csrf_token_to_cookie(request: Request, response: Response, token: str):
+    response.set_cookie(
+        COOKIE_NAME_CSRF_TOKEN,
+        value=token,
+        httponly=False,
+        secure=is_secure(),
+        samesite="Lax",
+        max_age=int(60 * dify_config.ACCESS_TOKEN_EXPIRE_MINUTES),
+        path="/",
+    )
+
+
+def _clear_cookie(
+    response: Response,
+    cookie_name: str,
+    samesite: str = "Lax",
+    http_only: bool = True,
+):
+    response.set_cookie(
+        cookie_name,
+        "",
+        expires=0,
+        path="/",
+        secure=is_secure(),
+        httponly=http_only,
+        samesite=samesite,
+    )
+
+
+def clear_access_token_from_cookie(response: Response, samesite: str = "Lax"):
+    _clear_cookie(response, COOKIE_NAME_ACCESS_TOKEN, samesite)
+
+
+def clear_refresh_token_from_cookie(response: Response):
+    _clear_cookie(response, COOKIE_NAME_REFRESH_TOKEN)
+
+
+def clear_csrf_token_from_cookie(response: Response):
+    _clear_cookie(response, COOKIE_NAME_CSRF_TOKEN, http_only=False)
+
+
+def check_csrf_token(request: Request):
+    # some apis are sent by beacon, so we need to bypass csrf token check
+    # since these APIs are post, they are already protected by SameSite: Lax, so csrf is not required.
+
+    for pattern in CSRF_WHITE_LIST:
+        if pattern.match(request.path):
+            return
+
+    csrf_token = extract_csrf_token(request)
+
+    def _unauthorized():
+        raise Unauthorized("CSRF token is missing or invalid.")
+
+    if not csrf_token:
+        _unauthorized()
+    verified = {}
+    try:
+        verified = PassportService().verify(csrf_token)
+    except:
+        _unauthorized()
+
+    exp: int | None = verified.get("exp")
+    if not exp:
+        _unauthorized()
+    else:
+        time_now = int(datetime.now().timestamp())
+        if exp < time_now:
+            _unauthorized()
+
+
+def generate_csrf_token() -> str:
+    exp_dt = datetime.now(UTC) + timedelta(minutes=dify_config.ACCESS_TOKEN_EXPIRE_MINUTES)
+    payload = {
```

Comment:
@wylswz The CSRF Token should be bound to a specific user, please add `user_id` here.

Reply:
done.